src/Microsoft.ML.Dnn/DnnUtils.cs

```diff
@@ -89,6 +90,24 @@ internal static Session LoadTFSession(IExceptionContext ectx, byte[] modelBytes,
             }
             return new Session(graph);
         }
+        internal static void MaybeDownloadFile(Uri address, string fileName)
+        {
+            using (WebClient client = new WebClient())
+            {
+                if (File.Exists(fileName))
+                {
+                    client.OpenRead(address);
+                    var totalSizeInBytes = Convert.ToInt64(client.ResponseHeaders["Content-Length"]);
+                    var currentSize = new FileInfo(fileName).Length;
+
+                    //If current file size is not equal to expected file size, re-download file
+                    if (currentSize != totalSizeInBytes)
+                        client.DownloadFile(new Uri($"{address}"), fileName);
+                }
```

Comment:
Lets write a test for this:
1) run code that downloads the meta file via running DNN training pipeline ... make sure it trains and you get expected accuracy.
2) modify the file by copying to new file such that you copy all bytes except for few so that size is different. Delete the original file and rename this file to the older file name so that next time DNN training runs the code tries to access the "incomplete" meta file and lines 97-106 get executed.
3) re-run this piece of code via running DNN training pipeline.

Reply:
I wrote a test that still will hit the line 105, but without step 1 to cut down on test time. 

---
In reply to: [333237915](https://github.com/dotnet/machinelearning/pull/4314#discussion_r333237915) [](ancestors = 333237915)

Reply:
Thanks, looks [great](https://codecov.io/gh/dotnet/machinelearning/pull/4314/diff#D1-102)